cpp/inc/bond/ext/grpc/detail/unary_call_impl.h

```diff
@@ -169,60 +170,68 @@ namespace bond { namespace ext { namespace gRPC { namespace detail
         // sent, regardless of whether there are any outstanding user
         // references still alive.
         std::atomic<size_t> _refCount{ 1 };
-
     };
 
-    /// @brief Detail class that helps implement \ref unary_call and \ref
-    /// shared_unary_call.
+
     template <typename Request, typename Response>
-    class unary_call_base
-    {
-        using impl_type = unary_call_impl<Request, Response>;
+    class unary_call_base;
 
-    public:
-        unary_call_base() = default;
+    template <typename Base>
+    class unary_call_impl_base;
 
-        explicit unary_call_base(boost::intrusive_ptr<impl_type> impl) noexcept
-            : _impl(std::move(impl))
-        {
-            BOOST_ASSERT(_impl);
-        }
+    template <typename Base>
+    class unary_call_input_base;
 
-        explicit operator bool() const noexcept
-        {
-            return static_cast<bool>(_impl);
-        }
+    template <typename Base>
+    class unary_call_result_base;
 
-        void swap(unary_call_base& rhs) noexcept
-        {
-            using std::swap;
-            swap(_impl, rhs._impl);
-        }
 
-        /// @brief Get the server context for this call.
-        const grpc::ServerContext& context() const noexcept
+    template <template <typename> class Base, typename Request, typename Response>
+    class unary_call_impl_base<Base<unary_call_base<Request, Response>>>
+    {
+    protected:
+        unary_call_base<Request, Response>& impl() noexcept
         {
-            return impl().context();
+            return static_cast<unary_call_base<Request, Response>&>(
+                static_cast<Base<unary_call_base<Request, Response>>&>(*this));
         }
 
-        /// @brief Get the server context for this call.
-        grpc::ServerContext& context() noexcept
+        const unary_call_base<Request, Response>& impl() const noexcept
         {
-            return impl().context();
+            return static_cast<const unary_call_base<Request, Response>&>(
+                static_cast<const Base<unary_call_base<Request, Response>>&>(*this));
         }
+    };
+
 
+    template <typename Request, typename Response>
+    class unary_call_input_base<unary_call_base<Request, Response>>
+        : public unary_call_impl_base<unary_call_input_base<unary_call_base<Request, Response>>>
+    {
+    public:
         /// @brief Get the request message for this call.
         const bonded<Request>& request() const noexcept
         {
-            return impl().request();
+            return this->impl().impl().request();
```

Comment:
`impl().impl()` is a code smell for me. Is there a way to unwrap the nesting? If these represent different things, are there more domain-specific names we can use?

Reply:
The first one returns the `unary_call_base` (CRTP) and then the second returns `unary_call_impl`. Didn't want to name the first one as `base`. Maybe something like `as_unary_call_base`?